### PR TITLE
Update OCCA Backend

### LINKED ISSUE #816: OCCA Backend
What is our path forward with the OCCA backend?

At this time, the OCCA backend is falling behind the other backends in terms of feature-completeness for preconditioner support. We have default functionality, but it pulls down to the CPU, so it would be slow for GPU environments. Also, we don't yet have GPU optimized or code-generation features it the OCCA backend. Third, currently an MFEM user can build the latest version of MFEM with OCCA support *or* with libCEED support but not both, as the libCEED OCCA backend uses a different version of OCCA than MFEM uses.

It totally makes sense that @dmed256 doesn't have time to keep this backend fully up to date, but I feel like we need someone with the time to keep this backend fully up to date. Currently I don't know enough C++ and OCCA to do the job.

I feel like we have a few options:

1) Retire OCCA backend - not my preferred option, as I feel like the OCCA backend could offer our first support on the new Intel GPUs

2) I learn OCCA and C++ - my preferred option, even though this will take a fair investment in time, I'm hoping what I learn will help our Rust GPU efforts

3) Invent more time for @dmed256 - not sure this is possible?

4) Other collaborators - I'm not sure if there are other collaborators who would be interested in working on this backend

I'm open to other ideas; I just want to make sure we have a plan here.

### PR DESCRIPTION
**Pull request by jeremylt**
Closes #816 

- [x] Update for changes in main (rebase or merge?)
- [ ] Move files used in JiT to `include/ceed/jit-source/occa` (I don't fully understand OCCA but I think we can skip this for now?)
- [x] Update `tests/junit.py` (replaces `tests/tap.sh` to only run tests for t1*-t3* on OCCA backends
- [x] Update `README.md` with new OCCA min version and backends
- [x] Update `docs/sphinx/source/releasenotes.md`
- [x] Update CI
- [x] Odd `OCCA_DIR/bin/occa` issue

### CODE REVIEW

**Comment by jeremylt** [comment]
```console
$ ENABLE_OPENCL="OFF" ./configure-cmake.sh
```
followed by
```
$ cmake --build build --parallel 16
```
on Noether leads to
```
[ 64%] Linking CXX executable bin/c/base
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clBuildProgram'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clEnqueueNDRangeKernel'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clSetKernelArg'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clReleaseMemObject'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clFinish'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clEnqueueUnmapMemObject'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clCreateSubBuffer'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clRetainCommandQueue'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clReleaseContext'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clCreateCommandQueueWithProperties'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clReleaseEvent'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clEnqueueWriteBuffer'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clEnqueueCopyBuffer'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clEnqueueReadBuffer'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clEnqueueBarrierWithWaitList'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetProgramBuildInfo'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clCreateContext'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clEnqueueMapBuffer'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetDeviceIDs'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetDeviceInfo'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clReleaseCommandQueue'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetPlatformIDs'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetPlatformInfo'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clWaitForEvents'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetEventProfilingInfo'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clCreateProgramWithBinary'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetKernelWorkGroupInfo'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clGetProgramInfo'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clCreateKernel'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clCreateBuffer'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clCreateProgramWithSource'
/usr/bin/ld: ../lib/libocca.so: undefined reference to `clReleaseKernel'
collect2: error: ld returned 1 exit status
gmake[2]: *** [tests/CMakeFiles/bin-c-base.dir/build.make:98: tests/bin/c/base] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1441: tests/CMakeFiles/bin-c-base.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

So it seems that disabling OpenCL is not being respected?

If I don't try to disable OpenCL, I get the warning
```
CMake Warning at CMakeLists.txt:82 (add_library):
  Cannot generate a safe runtime search path for target libocca because files
  in some directories may conflict with libraries in implicit directories:

    runtime library [libOpenCL.so.1] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/intel/oneapi/compiler/latest/linux/lib
      /opt/rocm/lib
```

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -22,20 +22,16 @@ noether-cpu:
...
     - echo "-------------- LIBXSMM -------------" && basename $XSMM_DIR
 # -- OCCA v1.1.0
-    - cd .. && export OCCA_VERSION=occa-1.1.0 OCCA_OPENCL_ENABLED=0 && { [[ -d $OCCA_VERSION ]] || { git clone --depth 1 --branch v1.1.0 https://github.com/libocca/occa.git $OCCA_VERSION && sed -i '/sysctl.h/d' $OCCA_VERSION/src/tools/sys.cpp && make -C $OCCA_VERSION -j$(nproc); }; } && export OCCA_DIR=$PWD/$OCCA_VERSION && cd libCEED
-    - echo "-------------- OCCA ----------------" && make -C $OCCA_DIR info
+    - cd .. && export OCCA_VERSION=occa-1.4.0 && { [[ -d $OCCA_VERSION ]] || { git clone --depth 1 --branch v1.4.0 https://github.com/libocca/occa.git $OCCA_VERSION && cd $OCCA_VERSION && export ENABLE_OPENCL="OFF" ENABLE_DPCPP="OFF" ENABLE_HIP="OFF" ENABLE_CUDA="OFF" && ./configure-cmake.sh && cmake --build build --parallel $NPROC_CPU && cmake --install build --prefix . && cd ..; }; } && export OCCA_DIR=$PWD/$OCCA_VERSION && cd libCEED
```
Comment (jeremylt) [inline .gitlab-ci.yml:25]: Something has to be a bit screwy here, as I'm getting
```
../occa/bin/occa: error while loading shared libraries: libocca.so: cannot open shared object file: No such file or directory
```
and only the `/cpu/self/occa` backend is showing as an OCCA mode

Comment (jedbrown) [reply .gitlab-ci.yml]: `readelf -d ../occa/bin/occa` should help explain. It looks like maybe RPATH wasn't set?

**Comment by jeremylt** [comment]
@kris-rowe, `clang-tidy` is generating this warning

```
/var/lib/gitlab-runner/builds/N8GCsqus/0/libceed/occa-1.4.0/install/include/occa/core/kernelArg.hpp:79:5: warning: Call to virtual method 'kernelArg::primitiveConstructor' during construction bypasses virtual dispatch [clang-analyzer-optin.cplusplus.VirtualCall]
    OCCA_GENERIC_CLASS_CONSTRUCTORS(kernelArg);
    ^
/var/lib/gitlab-runner/builds/N8GCsqus/0/libceed/occa-1.4.0/install/include/occa/types/generic.hpp:44:5: note: expanded from macro 'OCCA_GENERIC_CLASS_CONSTRUCTORS'
    primitiveConstructor(arg);                      \
```

should we be worried?

**Comment by kris-rowe** [comment]
I don't think the issues related to these warnings will cause any major problems. 

(Adding a linter to the OCCA CI pipeline is on the todo list. I am certain there are lots of things that will be uncovered when we do).

**Review by jedbrown — APPROVED** [review]
Thanks for this work. Is my understanding correct that this now runs the lower level tests in CI, and passes most higher level tests when using qfunctions (with the syntax workarounds) from a different OCCA branch that will be rebased after this merges? If so, that plan sounds good to me.

**Comment by jeremylt** [comment]
> Is my understanding correct that this now runs the lower level tests in CI, 

Yes, we test t0-t3 in CI

> and passes most higher level tests when using qfunctions (with the syntax workarounds) from a different OCCA branch that will be rebased after this merges?

That is my understanding as well

**Pull request merged**